File: src/main/java/com/github/svegon/capi/event/render/IsInvisibleToCallback.java
package com.github.svegon.capi.event.render;

import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;
import net.minecraft.entity.Entity;
import net.minecraft.entity.player.PlayerEntity;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;

public interface IsInvisibleToCallback {
    Event<IsInvisibleToCallback> EVENT = EventFactory.createArrayBacked(IsInvisibleToCallback.class,
            (entity, viewer, callback) -> {}, (listeners) -> (entity, viewer, callback) -> {
        for (IsInvisibleToCallback listener : listeners) {
            listener.invisibleSightCheck(entity, viewer, callback);

            if (callback.isCancelled()) {
                return;
            }
        }
            });

    void invisibleSightCheck(Entity entity, PlayerEntity viewer, CallbackInfoReturnable<Boolean> callback);
}

File: src/main/kotlin/io/github/svegon/mclientapi/event/render/IsInvisibleToCallback.kt
package io.github.svegon.capi.event.render

import io.github.svegon.capi.event.render.IsInvisibleToCallback
import net.fabricmc.fabric.api.event.Event
import net.fabricmc.fabric.api.event.EventFactory
import net.minecraft.entity.Entity
import net.minecraft.entity.player.PlayerEntity
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable
import java.util.function.Function
import kotlin.Array
import kotlin.Boolean
import kotlin.invoke

interface IsInvisibleToCallback {
    fun invisibleSightCheck(entity: Entity?, viewer: PlayerEntity?, callback: CallbackInfoReturnable<Boolean?>?)

    companion object {
        @JvmField
        val EVENT: Event<IsInvisibleToCallback?> = EventFactory.createArrayBacked(
            IsInvisibleToCallback::class.java,
            IsInvisibleToCallback { entity: Entity?, viewer: PlayerEntity?, callback: CallbackInfoReturnable<Boolean?>? -> },
            Function { listeners: Array<IsInvisibleToCallback?>? ->
                IsInvisibleToCallback { entity: Entity?, viewer: PlayerEntity?, callback: CallbackInfoReturnable<Boolean?>? ->
                    for (listener in listeners) {
                        listener.invisibleSightCheck(entity, viewer, callback)

                        if (callback.isCancelled()) {
                            return@IsInvisibleToCallback
                        }
                    }
                }
            })
    }
}
